### PR TITLE
feat: support allOf in a when context

### PR DESCRIPTION
**Pull request by garethjevans**

### CODE REVIEW

**Comment by MarkEWaite** [comment]
@nre-ableton could you review this?  @garethjevans has some downstream tests in Jenkins infra that would benefit from this PR

File: src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    def branch(String name) {
+        this.branches.add(name)
+    }
+
+    def expression(Closure closure) {
+        this.expressions.add(closure)
+    }
+
+    def expressions(delegate) {
+        def exp_result;
```
Comment (nre-ableton) [inline src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy:26]: Please remove this trailing semicolon.

File: src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    def branch(String name) {
+        this.branches.add(name)
+    }
+
+    def expression(Closure closure) {
+        this.expressions.add(closure)
+    }
+
+    def expressions(delegate) {
```
Comment (nre-ableton) [inline src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy:25]: This should be declared as `Object delegate`.

File: src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        def exp_result;
+        for (def exp in this.expressions) {
+            exp_result = executeWith(delegate, exp)
+            if (!exp_result) {
+                return false
+            }
```
Comment (nre-ableton) [inline src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy:31]: I would prefer:

```groovy
if (!executeWith(delegate, exp)) {
    return false
}
```

File: src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        }
+        return true
+    }
+
+    Boolean execute(Object delegate) {
```
Comment (nre-ableton) [inline src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy:48]: Please use the primitive `boolean` type here.

File: src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        AntPathMatcher antPathMatcher = new AntPathMatcher();
```
Comment (nre-ableton) [inline src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy:49]: Another trailing semicolon.

File: src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+        if (this.branches.size() > 0) {
+            for(def branch in this.branches) {
+                if (!antPathMatcher.match(branch, delegate.env.BRANCH_NAME) ) {
```
Comment (nre-ableton) [inline src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy:53]: Nit: tiny whitespace error at the end of this line. It should be:

```suggestion
                if (!antPathMatcher.match(branch, delegate.env.BRANCH_NAME)) {
```

File: src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+                    return false
+                }
+            }
+        }
+
+        if (this.expressions.size() > 0) {
+            boolean exp = expressions(delegate)
+            if (!exp) {
+                return false
+            }
```
Comment (nre-ableton) [inline src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy:63]: This can be made simpler:

```groovy
if (!expressions(delegate)) {
    return false
}
```

File: src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        }
+
+        return true
+    }
+}
```
Comment (nre-ableton) [inline src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy:75]: Nit: missing a newline at the end of the file

File: src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy [hunk 9]
```diff
@@ -26,6 +33,18 @@ class AllOfDeclaration extends WhenDeclaration {
         return true
     }
 
+    def anyOf(delegate) {
+        def exp_result;
+        for (def exp in this.anyOfs) {
+            exp_result = exp.execute(delegate)
+            //exp_result = execute(delegate, exp)
```
Comment (nre-ableton) [inline src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy:40]: Why is this line commented out? I suspect you meant to delete it. :slightly_smiling_face:

File: src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy [hunk 10]
```diff
@@ -26,6 +33,18 @@ class AllOfDeclaration extends WhenDeclaration {
         return true
     }
 
+    def anyOf(delegate) {
+        def exp_result;
```
Comment (nre-ableton) [inline src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy:37]: Another trailing semicolon.

File: src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy [hunk 11]
```diff
@@ -26,6 +33,18 @@ class AllOfDeclaration extends WhenDeclaration {
...
+        for (def exp in this.anyOfs) {
+            exp_result = exp.execute(delegate)
+            //exp_result = execute(delegate, exp)
+            if (!exp_result) {
+                return false
+            }
```
Comment (nre-ableton) [inline src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy:43]: Could be simpler:

```groovy
if (!exp.execute(delegate)) {
    return false
}
```

File: src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy [hunk 12]
```diff
@@ -44,6 +63,13 @@ class AllOfDeclaration extends WhenDeclaration {
             }
         }
 
+        if (this.anyOfs.size() > 0) {
+            boolean exp = anyOf(delegate)
+            if (!exp) {
+                return false
+            }
```
Comment (nre-ableton) [inline src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy:70]:
```groovy
if (!anyOf(delegate)) {
    return false
}
```

File: src/main/groovy/com/lesfurets/jenkins/unit/declarative/AnyOfDeclaration.groovy [hunk 13]
```diff
@@ -17,6 +20,22 @@ class AnyOfDeclaration extends WhenDeclaration {
...
 
+    def allOf(@DelegatesTo(strategy = DELEGATE_FIRST, value = AllOfDeclaration) Closure closure) {
+        this.allOfs.add(createComponent(AllOfDeclaration, closure))
+    }
+
+    def allOf(delegate) {
+        def exp_result;
+        for (def exp in this.allOfs) {
+            exp_result = exp.execute(delegate)
+            //exp_result = execute(delegate, exp)
```
Comment (nre-ableton) [inline src/main/groovy/com/lesfurets/jenkins/unit/declarative/AnyOfDeclaration.groovy:31]: Why is this line here?

File: src/main/groovy/com/lesfurets/jenkins/unit/declarative/AnyOfDeclaration.groovy [hunk 14]
```diff
@@ -17,6 +20,22 @@ class AnyOfDeclaration extends WhenDeclaration {
         this.expressions.add(closure)
     }
 
+    def allOf(@DelegatesTo(strategy = DELEGATE_FIRST, value = AllOfDeclaration) Closure closure) {
+        this.allOfs.add(createComponent(AllOfDeclaration, closure))
+    }
+
+    def allOf(delegate) {
+        def exp_result;
```
Comment (nre-ableton) [inline src/main/groovy/com/lesfurets/jenkins/unit/declarative/AnyOfDeclaration.groovy:28]: Trailing semicolon.

File: src/main/groovy/com/lesfurets/jenkins/unit/declarative/AnyOfDeclaration.groovy [hunk 15]
```diff
@@ -17,6 +20,22 @@ class AnyOfDeclaration extends WhenDeclaration {
...
+        for (def exp in this.allOfs) {
+            exp_result = exp.execute(delegate)
+            //exp_result = execute(delegate, exp)
+            if (!exp_result) {
+                return false
+            }
```
Comment (nre-ableton) [inline src/main/groovy/com/lesfurets/jenkins/unit/declarative/AnyOfDeclaration.groovy:34]:
```groovy
if (!exp.execute(delegate)) {
    return false
}
```

File: src/main/groovy/com/lesfurets/jenkins/unit/declarative/AnyOfDeclaration.groovy [hunk 16]
```diff
@@ -17,6 +20,22 @@ class AnyOfDeclaration extends WhenDeclaration {
         this.expressions.add(closure)
     }
 
+    def allOf(@DelegatesTo(strategy = DELEGATE_FIRST, value = AllOfDeclaration) Closure closure) {
+        this.allOfs.add(createComponent(AllOfDeclaration, closure))
+    }
+
+    def allOf(delegate) {
```
Comment (nre-ableton) [inline src/main/groovy/com/lesfurets/jenkins/unit/declarative/AnyOfDeclaration.groovy:27]: Should be declared as `Object delegate`.

File: src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy [hunk 17]
```diff
@@ -26,6 +33,18 @@ class AllOfDeclaration extends WhenDeclaration {
         return true
     }
 
+    def anyOf(delegate) {
```
Comment (nre-ableton) [inline src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy:36]: Should be declared as `Object delegate`.

**Review by nre-ableton — CHANGES_REQUESTED** [review]
I'm not terribly familiar with declarative pipelines, so I don't know if this is indeed the correct behavior for `anyOf`, but it seems right. :smile: Anyways, I have a few stylistic nitpicks, otherwise it looks alright.

**Comment by nre-ableton** [comment]
Note: the failing AppVeyor CI tests appears to be caused by a temporary network timeout and not failing tests etc.

File: src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy (hunk 4 above)
Comment (garethjevans) [reply src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy:33]: I can't change this without changing the return type of `WhenDeclaration.execute`

File: src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy (hunk 8 above)
Comment (garethjevans) [reply src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy:54]: copied from existing file.  If this is a requirement it should be fixed in a linter.

File: src/main/groovy/com/lesfurets/jenkins/unit/declarative/AnyOfDeclaration.groovy (hunk 15 above)
Comment (garethjevans) [reply src/main/groovy/com/lesfurets/jenkins/unit/declarative/AnyOfDeclaration.groovy]: reworked using any/every to simplify

File: src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy (hunk 3 above)
Comment (garethjevans) [reply src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy]: reworked using any/every to simplify

**Comment by garethjevans** [comment]
The main structure of `AllOfDeclaration` was taken from the existing code in `AnyOfDeclaration` which I've also fixed.  I didn't want to go too far down the route of updating `WhenDeclaration` and the rest to match the comments.

Using any/every closures simplifies the code quite nicely and is much more readable.

File: src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+    Boolean execute(Object delegate) {
+        def results = []
+
+        AntPathMatcher antPathMatcher = new AntPathMatcher()
+
+        if (this.branches.size() > 0) {
+            for(def branch in this.branches) {
+                results.add(antPathMatcher.match(branch, delegate.env.BRANCH_NAME))
+            }
+        }
```
Comment (nre-ableton) [inline src/main/groovy/com/lesfurets/jenkins/unit/declarative/AllOfDeclaration.groovy:42]: This could be simplified as such:

```groovy
branches.each { branch ->
    results.add(antPathMatcher.match(branch, delegate.env.BRANCH_NAME)) 
}
```

I realize that this was probably copied from AnyOfDeclaration. Still, I don't think that the implementations necessarily need to be identical.

**Review by nre-ableton — COMMENTED** [review]
Looks much better! :+1: Please squash all commits and I'll do a final review and merge.

**Review by nre-ableton — APPROVED** [review]
Thanks for the contribution!

**Comment by nre-ableton** [comment]
@timja I don't think I have the rights to do that. At any rate, I haven't made any releases yet. I would be open to doing this in the future but would need to be onboarded. Pinging @stchar, who has taken care of this in the past.

**Comment by timja** [comment]
You do have access:
https://github.com/jenkins-infra/repository-permissions-updater/blob/master/permissions/component-jenkins-pipeline-unit.yml#L10

These docs should cover it:
https://www.jenkins.io/doc/developer/publishing/releasing/

although it's probably a bit different as this project uses gradle, this might help:
https://github.com/jenkinsci/gradle-jpi-plugin

and the process in this repo:
https://github.com/jenkinsci/JenkinsPipelineUnit/blob/master/RELEASE.md

**Comment by nre-ableton** [comment]
@timja Ah, ok, I'll give it a shot then!

**Comment by nre-ableton** [comment]
@timja Ok, I made my first release; let's hope that I did everything correctly. :sweat_smile: :crossed_fingers: I published the release, but my understanding is that it might take the artifacts a little while to show up on the Artifactory repo, as you probably are aware.

**Comment by timja** [comment]
> @timja Ok, I made my first release; let's hope that I did everything correctly. 😅 🤞 I published the release, but my understanding is that it might take the artifacts a little while to show up on the Artifactory repo, as you probably are aware.

they are immediately available, I'm not sure what that comment in the release docs in this repo is referring to

i.e. see https://github.com/jenkins-infra/pipeline-library/pull/188 already available

**Comment by MarkEWaite** [comment]
There was a time when a significant delay existed between the publication of a Jenkins plugin and its visibility in the Jenkins update center.  That does not seem relevant with the component, since it is not a Jenkins plugin, but that may have been the source of the phrasing in the release docs.
